src/main/java/com/google/sps/servlets/utils/Statistics.java

```diff
@@ -14,29 +14,89 @@
   private Map<Range, Integer> aggregateValues;
   private double averageScore;
+  private static final double INTERVAL = 0.2;
+  private static final double UPPER_END = 1.0;
+  private static final double LOWER_END = -1.0;
+
+  public Map<Range, Integer> getAggregateValues() {
+    return aggregateValues;
+  }
+
+  public double getAverageScore() {
+    return averageScore;
+  }
 
   public Statistics(List<Double> sentimentScores) {
     setAggregateScores(sentimentScores);
     setAverageScore(sentimentScores);
   }
 
+  /**
+   * Categorize all score values into different range intervals and count the frequency for each
+   * interval, and set the aggregatedValues.
+   *
+   * @param sentimentScores a list of score values from -1.0 to 1.0
+   */
   private void setAggregateScores(List<Double> sentimentScores) {
-    // TODO(Xin): Add sorting code
     aggregateValues = new HashMap<>();
-    aggregateValues.put(new Range(-1.0, 1), 1);
+    BigDecimal curPoint = BigDecimal.valueOf(LOWER_END);
+    BigDecimal interval = BigDecimal.valueOf(INTERVAL);
+
+    // Initialize the HashMap with intervals
+    while (curPoint.doubleValue() < UPPER_END) {
+      Range currentRange =
+          new Range(
+              curPoint.doubleValue(), Math.min(curPoint.add(interval).doubleValue(), UPPER_END));
+      curPoint = curPoint.add(interval);
+      aggregateValues.put(currentRange, 0);
+    }
+
+    // Add score's interval to different ranges two sorting with and two pointers pop-up
+    sentimentScores.sort(Comparator.naturalOrder());
+    curPoint = BigDecimal.valueOf(LOWER_END);
+    int scoreIdx = 0;
+    while ((scoreIdx < sentimentScores.size()) && (curPoint.doubleValue() < UPPER_END)) {
+      double scoreVal = sentimentScores.get(scoreIdx);
+      // return null for edge cases
+      if ((scoreVal < LOWER_END) || (scoreVal > UPPER_END)) {
+        scoreIdx += 1;
+        continue;
+      }
+      // check available interval for scoreVal
+      BigDecimal nextPoint = curPoint.add(interval);
+      double intervalUpperVal = nextPoint.doubleValue();
+      if (scoreVal < intervalUpperVal || intervalUpperVal == UPPER_END) {
+        Range foundRange = new Range(curPoint.doubleValue(), intervalUpperVal);
+        aggregateValues.put(foundRange, aggregateValues.get(foundRange) + 1);
+        scoreIdx += 1;
+      } else {
+        curPoint = nextPoint;
+      }
+    }
   }
 
+  /**
+   * Set the average score of given sentiment scores. Returns -99 if the average score is not valid
+   * or none of the sentiment scores is valid.
+   *
+   * @param sentimentScores a list of score values from -1.0 to 1.0
+   */
   private void setAverageScore(List<Double> sentimentScores) {
-    // TODO(Xin): Add average score method
-    averageScore = 0;
+    averageScore =
+        sentimentScores.stream()
+            .mapToDouble(i -> i)
+            .filter(score -> (score >= LOWER_END && score <= UPPER_END))
```

Comment:
Do you need this filter since the API will not give you anything outside of the range?
Also, you have to call setAggregateScores before this, where you check the values already in line 71. I suggest removing.

Reply:
Solved by filtering at the beginning of constructor